### PR TITLE
Cache GitHub requests

### PR DESCRIPTION
**Pull request by CyberShadow**
Cache requests to the GitHub API as recommended by
https://developer.github.com/v3/#conditional-requests .

Avoids using up rate-limit points and re-downloading the same data for
repeated requests.

Data is never stale; cache is used only when the server claims nothing
has changed since our last request.

Requires https://github.com/lukechilds/cacheable-request/pull/77.

### CODE REVIEW

File: lib/platform/github/gh-got-wrapper.js [hunk 1]
```diff
@@ -19,7 +33,11 @@ async function get(path, options, retries = 5) {
   opts.baseUrl = opts.endpoint;
   delete opts.endpoint;
   const method = opts.method || 'get';
-  const useCache = opts.useCache || true;
+  const useCache = opts.useCache || true; // in-memory cache
+  opts.cache = cacheMap; /* persistent cache - always validated, but
+                            hits don't count against request limit */
+  opts.shared = false; // https://github.com/kornelski/http-cache-semantics/issues/9
```
Comment (rarkins) [inline lib/platform/github/gh-got-wrapper.js]: Does this mean that caching is done uniquely based on the authentication token used, for example?

File: lib/platform/github/gh-got-wrapper.js [hunk 2]
```diff
@@ -19,7 +33,11 @@ async function get(path, options, retries = 5) {
...
+  opts.forceRefresh = true;
```
Comment (rarkins) [inline lib/platform/github/gh-got-wrapper.js]: It would be good to document this too

**Review by rarkins — CHANGES_REQUESTED** [review]
I would definitely like to make this feature optional, and maybe disabled by default to begin with until others have opted in to test it at scale first.

The last time I enabled this form of caching in Renovate it was catastrophic in the app.. although I suspect *most* of that problem was the empty 304 body problem you are trying to fix already.

File: lib/platform/github/gh-got-wrapper.js (hunk 1 above)
Comment (CyberShadow) [reply lib/platform/github/gh-got-wrapper.js]: My understanding of this is that RFC 7234 forbids caching in this situation as an additional precaution against serving private cached data to unauthenticated requests. However, servers should already specify the header used for authentication in the `Vary` response header (at least GitHub does), and, assuming the cache driver implements this mechanism correctly, this "should" work as intended.

**Comment by CyberShadow** [comment]
> I would definitely like to make this feature optional, and maybe disabled by default to begin with until others have opted in to test it at scale first.

> The last time I enabled this form of caching in Renovate it was catastrophic in the app.. although I suspect most of that problem was the empty 304 body problem you are trying to fix already.

In theory, this could be seen as an optimization / "bugfix" (in that renovate is not following the server and the API documentation's request to cache responses), and wastefully consumes rate limit points; as all entries are revalidated, it should never result in stale data, and the patch should have no effect other than an optimization.

In practice, a lot of things can go wrong. Bugs throughout the stack used by Renovate (such as https://github.com/lukechilds/cacheable-request/pull/77), cache misconfiguration (such as possibly the question raised above), or bugs in API endpoints' implementations (I've had GitHub API reply with "304 Not Modified" when a certain subresource included in the response was actually modified in at least one circumstance) may introduce bad behavior.

I implemented this patch because I'm "self-hosting" Renovate and using it from my personal GitHub account, which shares its API rate limit with other things running on it as well. However, if this patch will not help Renovate users in any significant use case, then perhaps it's better to let sleeping dogs lie.

**Comment by rarkins** [comment]
I definitely think this would be a good option to allow self-hosters to opt into. There would be many cases where data is unchanged but meanwhile the cache signature (e.g. token used) would remain the same so the cache would be in use.

For the app, it would be of minimal benefit, because:
- The majority of queries are repo-related, i.e specific to each repo and not shared with others even if the same owner
- We memory cache all GET requests for the duration of a repo run, i.e. we don't even check for unchanged
- GitHub forces rotation of tokens regularly - we do it twice per hour per account - so it's not often that the same token is reused between runs because it's not common that the same repo is run twice closely together

Also, ensuring we completely clear our (memory) cache between jobs also reduces the chance we accidentally leak data between users by accident. So to get cache benefit across runs we'd need to swap the cache in and out from a cold storage for each run.

**Comment by CyberShadow** [comment]
> I would definitely like to make this feature optional

Sorry, I can't figure out how to get to Renovate's config from `gh-got-wrapper.js`.

> Also, ensuring we completely clear our (memory) cache between jobs also reduces the chance we accidentally leak data between users by accident.

Another way to achieve this without actually discarding the data is to make the user ID or (hash of) the authentication token part of the cache namespace.

**Comment by rarkins** [comment]
> Sorry, I can't figure out how to get to Renovate's config from gh-got-wrapper.js.

Let's keep it simple right now and just look for `process.env.RENOVATE_HTTP_CACHING`. If it's set, then enable the caching in the github wrapper.

> Another way to achieve this without actually discarding the data is to make the user ID or (hash of) the authentication token part of the cache namespace.

For me, it's still not a safe enough partitioning between the users when it comes to the app.

File: lib/platform/github/gh-got-wrapper.js [hunk 3]
```diff
@@ -19,7 +33,13 @@ async function get(path, options, retries = 5) {
   opts.baseUrl = opts.endpoint;
   delete opts.endpoint;
   const method = opts.method || 'get';
-  const useCache = opts.useCache || true;
+  const useCache = opts.useCache || true; // in-memory cache
```
Comment (ViceIce) [inline lib/platform/github/gh-got-wrapper.js]: This line is always true!

Comment (ViceIce) [reply lib/platform/github/gh-got-wrapper.js]: so we always use cache
https://github.com/renovatebot/renovate/blob/45dba15b7464ee30fb76f2747a2909b49d3578f2/lib/platform/github/gh-got-wrapper.js#L27

Comment (rarkins) [reply lib/platform/github/gh-got-wrapper.js]: Whoops. perhaps should be `opts.useCache !== false`

Comment (ViceIce) [reply lib/platform/github/gh-got-wrapper.js]: https://github.com/renovatebot/renovate/blob/master/lib/platform/bitbucket-server/bb-got-wrapper.js#L19-L20

**Comment by rarkins** [comment]
I think we want to use cache by default unless disabled explicitly?

**Comment by rarkins** [comment]
Closing this for now, because I would like to implement a more generic and configurable solution. Requirements to be discussed here: #3829

**Pull request closed without merging**